c++/src/kj/string.h

```diff
@@ -133,11 +133,15 @@ class StringPtr {
   inline constexpr const char* end() const { return content.end() - 1; }
 
   inline constexpr bool operator==(decltype(nullptr)) const { return content.size() <= 1; }
+  inline constexpr bool operator<=(decltype(nullptr)) const { return *this == nullptr; }
+  inline constexpr bool operator>=(decltype(nullptr)) const { return true; }
+  inline constexpr bool operator< (decltype(nullptr)) const { return false; }
+  inline constexpr bool operator> (decltype(nullptr)) const { return content.size() > 1; }
 
-  inline bool operator==(const StringPtr& other) const;
-  inline bool operator< (const StringPtr& other) const;
+  inline bool operator==(const StringPtr& other) const { return content.first(content.size() - 1) == other.content.first(other.content.size() - 1); }
```

Comment:
why -1 here but not below? If this is expected, then it needs careful commenting.

Reply:
We decided to only prserve legacy behavior in this pr, even though we both agreed that this behavior is wrong.

Reply:
Let's open a seprate PR with a test and a fix. Looks to me current behavior is very broken.

Reply:
I added a comment, I think the behavior is actually right. Basically, `StringPtr` is a `NUL` terminated slice. Thus, the last character in the `content` array does not need to be compared.
However, the same length should not be used for operator< because it enables a faster short circuit.

Reply:
> Looks to me current behavior is very broken.

How so? In both the cases of `operator==` and `operator<`, including or not including the `-1` makes no difference to the final outcome, because the last character is always NUL.

> enables a faster short circuit.

TBH I doubt this matters in practice.

Reply:
Originally I thought the StringPtr could be sliced without being NUL terminated and that's where my confusion stemmed from.